### DISCUSSION TITLE
Access Expo SecureStore from custom native code

### DISCUSSION THREAD

**Discussion by alexanderdavide** [expo/expo#18289] [Archived: Questions] [answered]
I'm using Expo's [SecureStore](https://docs.expo.dev/versions/latest/sdk/securestore/) to securely store a user identification respectively API key.

The app also sets up an Android Service that subscribes to the ACTION_SCREEN_ON intent. In this Service I need to access the API key that I've saved with SecureStore to perform authenticated API calls. What I've tried:

* As described by the docs, SecureStore on Android uses SharedPreferences. [This way](https://developer.android.com/training/data-storage/shared-preferences#GetSharedPreferences) I am able to get a value for the token from the default SharedPreferences, however it's encrypted as it should be.
* I tried to use the implementation of SecureStore because I thought all the logic would already be there to get me the decrypted value. However, I'm unable to successfully use the module. `getValueWithKeyAsync` expects an implementation of `ReadableArguments` which operates on a `Collection<String>` and I don't understand where to get it from. Passing a new Collection, of course, misses relevant data.

```java
@Override
public void onReceive(Context context, Intent intent) {
  String action = intent.getAction();
  if(!Intent.ACTION_SCREEN_ON.equals(intent.getAction())) return;
  SecureStoreModule secureStoreModule = new SecureStoreModule(context);
  Promise tokenPromise = new Promise() {
    @Override
    public void resolve(Object value) {
      // Perform authenticated API call
    }

    @Override
    public void reject(String code, String message, Throwable e) { 
      ...
    }
  };
  ReadableArguments readableArguments = new ReadableArguments() {
    // How to correctly implement this or where to get the correct implementation from?
  }
  secureStoreModule.getValueWithKeyAsync("TOKEN", readableArguments, tokenPromise);
}
```

I'd appreciate anything guiding me in the right direction. If you can think of other solutions for secure storage in React Native and accessing this secure storage in custom native code, I'm open for suggestions.

Besides, I've too posted this question on [StackOverflow](https://stackoverflow.com/questions/73020969/access-expo-securestore-from-custom-native-code).

**Comment by barthap** [comment, marked as answer]
I'm pretty sure there's another way of accessing React Native / Expo modules from custom native code, instead of creating the new instance, like you do now. But your way should eventually work too, but I don't like duplication of the module instance.

Do you want to use it from inside another module/package (is it react-native module or expo-module then)? or rather directly from the _app_ project?

Regarding `ReadableArguments` - IIRC it's just an interface. Instead of creating a custom implementation, you should use methods from the[ `Arguments` class](https://github.com/facebook/react-native/blob/main/ReactAndroid/src/main/java/com/facebook/react/bridge/Arguments.java).

**Reply by alexanderdavide** [reply]
Thanks for pointing out the `Arguments` class.

Actually, I hope there is a way to access the instance that must be existing somewhere in the runtime. I'd assume it to be the easiest solution but I couldn't figure it out or find a guide online.

I want to access it directly from my app project. It's not a package or similar.

For the time being, I've solved my problem with EncryptedSharedPreferences. While this works, it is related to more code and maintenance on my end. I'd still be happy to hand this over to Expo's SecureStore.

**Reply by barthap** [reply]
OK, I dig down a little bit and this should be possible without (much) hacking.
If you're inside React Native activity, the Android `context` should be an instance of `ReactApplicationContext`. (alternatively, try hacking `MainActivity.java` for other possible ways of retrieving `ReactContext`)
From the RN context, you can get RN native module called `NativeModulesProxy`, which is a proxy for all Expo modules. This API is very limited/cumbersome from native side, but from there, you can call any module method. So the code is sth like:
```java
      Bundle options = new Bundle();
      // you can set options here using android Bundle
      
      List<Object> args = new ArrayList<>();
      args.add("TOKEN");
      args.add(options); // alternatively try args.add(Arguments.makeNativeMap(options));
                                      // because I'm not sure here

      Promise promise = new com.facebook.react.bridge.PromiseImpl(result -> {
        // result is a vararg of Objects, not sure what it really returns
        // success logic here
      }, error -> {
        // error logic here
      });
      
      ReactApplicationContext ctx = (ReactApplicationContext)context;
      NativeModulesProxy proxy = ctx.getNativeModule(NativeModulesProxy.class);
      assert proxy != null;
      proxy.callMethod("ExpoSecureStore",
        new DynamicFromObject("getValueWithKeyAsync"),
        Arguments.fromList(args),
        promise);
```

**Reply by alexanderdavide** [reply]
Thanks a lot for the help. I appreciate it!

Having the need to access `ReactApplicationContext` indeed makes it a bit impractical as it is seldom to "naturally" have access to it.

I'll look into it more as soon as I have the possibility to conduct refactorings.

**Reply by xuanmd** [reply]
Hi there, do we have any official way to access Secure Store from native code? My purpose is to get credentials and make API requests on an Android widget.
